### ISSUE TITLE
Can't cross compile

### ISSUE THREAD

**Issue by pauldotknopf** [Intel-Media-SDK/MediaSDK#94]
In the cmake files, ```CMAKE_C_FLAGS``` are being overwritten.

https://github.com/Intel-Media-SDK/MediaSDK/blob/e54022c1dbdcbd93b502f0da37352c9e9d103180/builder/FindGlobals.cmake#L95

My cross compile sets up these values for cmake using a ```toolchain.cmake``` file.

```
set( CMAKE_SYSTEM_NAME Linux )
set( CMAKE_SYSTEM_PROCESSOR x86_64 )
set( CMAKE_C_COMPILER x86_64-poky-linux-gcc )
set( CMAKE_CXX_COMPILER x86_64-poky-linux-g++ )
set( CMAKE_ASM_COMPILER x86_64-poky-linux-gcc )
set( CMAKE_AR x86_64-poky-linux-ar CACHE FILEPATH "Archiver" )
set( CMAKE_C_FLAGS " -m64 -march=corei7 -mtune=corei7 -mfpmath=sse -msse4.2  --sysroot=/home/pknopf/git/x3/abrarecipes/build/tmp/work/corei7-64-poky-linux/msdk/git-r0/recipe-sysroot  -O2 -pipe -g -feliminate-unused-debug-types -fdebug-prefix-map=/home/pknopf/git/x3/abrarecipes/build/tmp/work/corei7-64-poky-linux/msdk/git-r0=/usr/src/debug/msdk/git-r0 -fdebug-prefix-map=/home/pknopf/git/x3/abrarecipes/build/tmp/work/corei7-64-poky-linux/msdk/git-r0/recipe-sysroot-native= -fdebug-prefix-map=/home/pknopf/git/x3/abrarecipes/build/tmp/work/corei7-64-poky-linux/msdk/git-r0/recipe-sysroot=   -m64 -march=corei7 -mtune=corei7 -mfpmath=sse -msse4.2  --sysroot=/home/pknopf/git/x3/abrarecipes/build/tmp/work/corei7-64-poky-linux/msdk/git-r0/recipe-sysroot" CACHE STRING "CFLAGS" )
set( CMAKE_CXX_FLAGS " -m64 -march=corei7 -mtune=corei7 -mfpmath=sse -msse4.2  --sysroot=/home/pknopf/git/x3/abrarecipes/build/tmp/work/corei7-64-poky-linux/msdk/git-r0/recipe-sysroot  -O2 -pipe -g -feliminate-unused-debug-types -fdebug-prefix-map=/home/pknopf/git/x3/abrarecipes/build/tmp/work/corei7-64-poky-linux/msdk/git-r0=/usr/src/debug/msdk/git-r0 -fdebug-prefix-map=/home/pknopf/git/x3/abrarecipes/build/tmp/work/corei7-64-poky-linux/msdk/git-r0/recipe-sysroot-native= -fdebug-prefix-map=/home/pknopf/git/x3/abrarecipes/build/tmp/work/corei7-64-poky-linux/msdk/git-r0/recipe-sysroot=  -fvisibility-inlines-hidden  -m64 -march=corei7 -mtune=corei7 -mfpmath=sse -msse4.2  --sysroot=/home/pknopf/git/x3/abrarecipes/build/tmp/work/corei7-64-poky-linux/msdk/git-r0/recipe-sysroot" CACHE STRING "CXXFLAGS" )
set( CMAKE_ASM_FLAGS " -m64 -march=corei7 -mtune=corei7 -mfpmath=sse -msse4.2  --sysroot=/home/pknopf/git/x3/abrarecipes/build/tmp/work/corei7-64-poky-linux/msdk/git-r0/recipe-sysroot  -O2 -pipe -g -feliminate-unused-debug-types -fdebug-prefix-map=/home/pknopf/git/x3/abrarecipes/build/tmp/work/corei7-64-poky-linux/msdk/git-r0=/usr/src/debug/msdk/git-r0 -fdebug-prefix-map=/home/pknopf/git/x3/abrarecipes/build/tmp/work/corei7-64-poky-linux/msdk/git-r0/recipe-sysroot-native= -fdebug-prefix-map=/home/pknopf/git/x3/abrarecipes/build/tmp/work/corei7-64-poky-linux/msdk/git-r0/recipe-sysroot=   -m64 -march=corei7 -mtune=corei7 -mfpmath=sse -msse4.2  --sysroot=/home/pknopf/git/x3/abrarecipes/build/tmp/work/corei7-64-poky-linux/msdk/git-r0/recipe-sysroot" CACHE STRING "ASM FLAGS" )
set( CMAKE_C_FLAGS_RELEASE "-DNDEBUG" CACHE STRING "Additional CFLAGS for release" )
set( CMAKE_CXX_FLAGS_RELEASE "-DNDEBUG" CACHE STRING "Additional CXXFLAGS for release" )
set( CMAKE_ASM_FLAGS_RELEASE "-DNDEBUG" CACHE STRING "Additional ASM FLAGS for release" )
set( CMAKE_C_LINK_FLAGS " -m64 -march=corei7 -mtune=corei7 -mfpmath=sse -msse4.2  --sysroot=/home/pknopf/git/x3/abrarecipes/build/tmp/work/corei7-64-poky-linux/msdk/git-r0/recipe-sysroot  -Wl,-O1 -Wl,--hash-style=gnu -Wl,--as-needed" CACHE STRING "LDFLAGS" )
set( CMAKE_CXX_LINK_FLAGS " -m64 -march=corei7 -mtune=corei7 -mfpmath=sse -msse4.2  --sysroot=/home/pknopf/git/x3/abrarecipes/build/tmp/work/corei7-64-poky-linux/msdk/git-r0/recipe-sysroot  -O2 -pipe -g -feliminate-unused-debug-types -fdebug-prefix-map=/home/pknopf/git/x3/abrarecipes/build/tmp/work/corei7-64-poky-linux/msdk/git-r0=/usr/src/debug/msdk/git-r0 -fdebug-prefix-map=/home/pknopf/git/x3/abrarecipes/build/tmp/work/corei7-64-poky-linux/msdk/git-r0/recipe-sysroot-native= -fdebug-prefix-map=/home/pknopf/git/x3/abrarecipes/build/tmp/work/corei7-64-poky-linux/msdk/git-r0/recipe-sysroot=  -fvisibility-inlines-hidden  -m64 -march=corei7 -mtune=corei7 -mfpmath=sse -msse4.2  --sysroot=/home/pknopf/git/x3/abrarecipes/build/tmp/work/corei7-64-poky-linux/msdk/git-r0/recipe-sysroot -Wl,-O1 -Wl,--hash-style=gnu -Wl,--as-needed" CACHE STRING "LDFLAGS" )

# only search in the paths provided so cmake doesnt pick
# up libraries and tools from the native build machine
set( CMAKE_FIND_ROOT_PATH /home/pknopf/git/x3/abrarecipes/build/tmp/work/corei7-64-poky-linux/msdk/git-r0/recipe-sysroot /home/pknopf/git/x3/abrarecipes/build/tmp/work/corei7-64-poky-linux/msdk/git-r0/recipe-sysroot-native  /home/pknopf/git/x3/abrarecipes/build/tmp/work/corei7-64-poky-linux/msdk/git-r0/recipe-sysroot-native/usr/bin/perl-native  /home/pknopf/git/x3/abrarecipes/build/tmp/work/corei7-64-poky-linux/msdk/git-r0/git )
set( CMAKE_FIND_ROOT_PATH_MODE_PACKAGE ONLY )
set( CMAKE_FIND_ROOT_PATH_MODE_PROGRAM ONLY )
set( CMAKE_FIND_ROOT_PATH_MODE_LIBRARY ONLY )
set( CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ONLY )

# Use qt.conf settings
set( ENV{QT_CONF_PATH} /home/pknopf/git/x3/abrarecipes/build/tmp/work/corei7-64-poky-linux/msdk/git-r0/qt.conf )

# We need to set the rpath to the correct directory as cmake does not provide any
# directory as rpath by default
set( CMAKE_INSTALL_RPATH  )

# Use native cmake modules
list(APPEND CMAKE_MODULE_PATH "/home/pknopf/git/x3/abrarecipes/build/tmp/work/corei7-64-poky-linux/msdk/git-r0/recipe-sysroot/usr/share/cmake/Modules/")

# add for non /usr/lib libdir, e.g. /usr/lib64
set( CMAKE_LIBRARY_PATH /usr/lib /lib)
```

I pass ```--cross=toolchain.cmake``` to the ```build_mfx.pl``` script.

What is the purpose of providing the ```toolchain.cmake``` file if the values just get overwritten?

**Comment by pauldotknopf** [comment]
Things like this are happening in the ```media-driver`` repository as well.

https://github.com/intel/media-driver/issues/108

**Comment by DavidMarec** [comment]
To enable a build with Yocto there are some patch to apply.

```diff
--- a/builder/FindGlobals.cmake                                                                                                                                                                                     
+++ b/builder/FindGlobals.cmake
@@ -94,15 +94,14 @@

   set(warnings "-Wall -Wformat -Wformat-security -Wnon-virtual-dtor")

-  set(CMAKE_C_FLAGS   "-pipe -fPIC ${warnings} ${no_warnings}")
-  set(CMAKE_CXX_FLAGS "-pipe -fPIC ${warnings} ${no_warnings}")
+  set(CMAKE_POSITION_INDEPENDENT_CODE ON)
   append("-fPIE -pie" CMAKE_EXEC_LINKER_FLAGS)

   # CACHE + FORCE should be used only here to make sure that this parameters applied globally
-  set(CMAKE_C_FLAGS_DEBUG     "-O0 -g -D_DEBUG" CACHE STRING "" FORCE)
-  set(CMAKE_C_FLAGS_RELEASE   "-O2 -D_FORTIFY_SOURCE=2 -fstack-protector -DNDEBUG"    CACHE STRING "" FORCE)
-  set(CMAKE_CXX_FLAGS_DEBUG   "-O0 -g -D_DEBUG" CACHE STRING "" FORCE)
-  set(CMAKE_CXX_FLAGS_RELEASE "-O2 -D_FORTIFY_SOURCE=2 -fstack-protector -DNDEBUG"    CACHE STRING "" FORCE)
+  set(CMAKE_C_FLAGS_DEBUG     "${CMAKE_C_FLAGS_DEBUG} -Wall ${no_warnings}  -Wformat -Wformat-security -g -D_DEBUG" CACHE STRING "" FORCE)
+  set(CMAKE_C_FLAGS_RELEASE   "${CMAKE_C_FLAGS_RELEASE} -O2 -D_FORTIFY_SOURCE=2 -fstack-protector -Wall ${no_warnings} -Wformat -Wformat-security -DNDEBUG"    CACHE STRING "" FORCE)
+  set(CMAKE_CXX_FLAGS_DEBUG   "${CMAKE_CXX_FLAGS_DEBUG} -Wall ${no_warnings}  -Wformat -Wformat-security -g -D_DEBUG" CACHE STRING "" FORCE)
+  set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -O2 -D_FORTIFY_SOURCE=2 -fstack-protector -Wall ${no_warnings}  -Wformat -Wformat-security -DNDEBUG"    CACHE STRING "" FORCE)

   if ( Darwin )
     if (CMAKE_C_COMPILER MATCHES clang)
--- a/api/opensource/mfx_dispatch/CMakeLists.txt                                                                                                                                                                    
+++ b/api/opensource/mfx_dispatch/CMakeLists.txt
@@ -86,13 +86,14 @@

   set(no_warnings "-Wno-unknown-pragmas -Wno-unused")

-  set(CMAKE_CXX_FLAGS "-pipe -fPIC -std=c++11")
-  set(CMAKE_CXX_FLAGS_DEBUG   "-O0 -Wall ${no_warnings} -Wformat -Wformat-security -g -D_DEBUG")
-  set(CMAKE_CXX_FLAGS_RELEASE "-O2 -D_FORTIFY_SOURCE=2 -fstack-protector -Wall ${no_warnings} -Wformat -Wformat-security  -DNDEBUG")
+  set(CMAKE_POSITION_INDEPENDENT_CODE ON)
+  set (CMAKE_CXX_STANDARD 11)
+  set(CMAKE_CXX_FLAGS_DEBUG   "${CMAKE_CXX_FLAGS_DEBUG} -O0 -Wall ${no_warnings} -Wformat -Wformat-security -g")
+  set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -O2 -D_FORTIFY_SOURCE=2 -fstack-protector -Wall ${no_warnings} -Wformat -Wformat-security")

-  if (DEFINED CMAKE_FIND_ROOT_PATH)
-    append("--sysroot=${CMAKE_FIND_ROOT_PATH} " LINK_FLAGS)
-  endif (DEFINED CMAKE_FIND_ROOT_PATH)
+#  if (DEFINED CMAKE_FIND_ROOT_PATH)
+#    append("--sysroot=${CMAKE_FIND_ROOT_PATH} " LINK_FLAGS)
+#  endif (DEFINED CMAKE_FIND_ROOT_PATH)
 endif( )

 include_directories (
```

IMO, CMake files need a serious code review to fully allow cross-compiling.

**Comment by alariono** [comment]
@DavidMarec,
Thanks for you finding!
Could you summarize what would you like to change?

**Comment by DavidMarec** [comment]
For example, 

* Do not overwrite any `CMAKE_xxx` internal variable:
* do not presume of the compiler id :
```cmake 
if ( Darwin )
     if (CMAKE_C_COMPILER MATCHES clang) 
```
CMAKE is supposed to do this job already.
* use default CMAKE built-in options (and computed for a given compiler) like `CMAKE_POSITION_INDEPENDENT_CODE` and `CMAKE_CXX_STANDARD`.

* The toolchain for Cross compilation must be done in a separate toolchain.cmake, given by the operator, this is what paul pointed at, do not care about it in your own CMake files.

The patch above is working if running Yocto-rocko, but I do not have checked for any other compiler or other building system.

I don't known the meaning and history that lays behind the options that were appended or removed, so It's difficult to just say what is relevant or not.

**Comment by dvrogozh** [comment]
@DavidMarec : could you, please, take a look on the changes just landed into the mediasdk repository to avoid compiler flags overrides (see above)? We would like to know whether you have anything else which results in the build failures for you trying to cross compile? If there is anything, we will prioritize work on these items.

Are these 2 below critical for what you need? As I mentioned in the cross-compile PR fixing these will require change of cmake requirement... We will need some time to see this happen...
```
set(CMAKE_CXX_STANDARD 11)
set(CMAKE_POSITION_INDEPENDENT_CODE ON)
```

**Comment by DavidMarec** [comment]
@dvrogozh, I removed my patches and setup the yocto recipe to the last revision and the build successfully ends up.
So, this part sounds ok, thanks.

The CMAKE statements you pointed at are not fully required, for the moment. The purpose of those macros is to be the more compiler-agnostic as possible.
_"-std=" and "fpic/fpie" are GCC standards and PICs macros have no signification if building for windows if I remember well_

CMake version requirement have to be bumped to 3.1 to support both.
_2.8 for pic/pie_

**Comment by dvrogozh** [comment]
@DavidMarec : so, we can cross compile now. Are you ok to close this issue and submit others when they spark blocking cross compilation?

**Comment by DavidMarec** [comment]
@dvrogozh 
from my point of view, it s ok.
But, I m  not the one who opened the issue.

**Comment by dvrogozh** [comment]
Ups. @pauldotknopf : are you ok to close the issue?

**Comment by pauldotknopf** [comment]
I can't test right now, but it looks good to me.

I will open another issue if there is anything else.

Thanks!

**Comment by dvrogozh** [comment]
Sound good. Thank you.
